src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -1651,79 +1657,158 @@ size_t runtime::gpu::CUDAEmitter::build_softmax(const std::vector<std::string>&
+            CudaKernelBuilder::add_pod_typedefs(writer);
+            runtime::gpu::CudaKernelBuilder::get_softmax_op(
+                writer, kernel_name, args, dtypes_str, non_reduce_rank, reduce_rank);
+            compiled_kernel = m_ctx->compiled_kernel_pool->set(kernel_name, writer.get_code());
         }
-    }
-    NVShape output_strides = row_major_strides(output_shape);
-    uint32_t nthreads = static_cast<uint32_t>(shape_size(output_shape));
-    // TODO: currently we set it to 64, will add tuning method later
-    uint32_t block_size_x = 64;
-    if (reduce_flag.back() == 1)
-    {
-        block_size_x = 8;
-    }
-    uint32_t aligned_grid_size_x = align_to_block_size(nthreads, block_size_x);
-    auto args = m_primitive_emitter->add_kernel_args();
-    args.add_placeholder(dtypes[0], "in")
-        .add_placeholder(dtypes[1], "out")
-        .add("out_strides", output_strides)
-        .add("non_reduce_strides", non_reduce_strides)
-        .add("reduce_shape", reduce_shape)
-        .add("reduce_strides", reduce_strides)
-        .add("nthreads", nthreads);
 
-    // if the kernel has not been compiled, build it
-    auto compiled_kernel = m_ctx->compiled_kernel_pool->get(kernel_name);
-    if (compiled_kernel == nullptr)
-    {
-        codegen::CodeWriter writer;
-        CudaKernelBuilder::add_pod_typedefs(writer);
-        runtime::gpu::CudaKernelBuilder::get_softmax_op(
-            writer, kernel_name, args, dtypes, non_reduce_rank, reduce_rank);
-        compiled_kernel = m_ctx->compiled_kernel_pool->set(kernel_name, writer.get_code());
+        std::unique_ptr<gpu::primitive> softmax(
+            new gpu::primitive{[=](void** inputs, void** outputs) mutable {
+                void** args_list = args.resolve_placeholder(0, &inputs[0])
+                                       .resolve_placeholder(1, &outputs[0])
+                                       .get_argument_list();
+
+                CUDA_SAFE_CALL(cuLaunchKernel(*compiled_kernel.get(),
+                                              aligned_grid_size_x,
+                                              1,
+                                              1,
+                                              block_size_x,
+                                              1,
+                                              1,
+                                              0,
+                                              nullptr,
+                                              args_list,
+                                              nullptr));
+                debug_sync();
+            }});
+
+        return this->m_primitive_emitter->register_primitive(softmax, hash);
     }
+    else
+    {
+        uint32_t reduce_count = static_cast<uint32_t>(shape_size(reduce_shape));
+        uint32_t block_size_x = 1;
+        while ((block_size_x << 1) <= reduce_count)
+        {
+            block_size_x <<= 1;
+        }
+        block_size_x = fmin(512, block_size_x);
```

Comment:
Infinite loop if reduce_count > 0x80000000.
This might be better done something like this:
```
uint32_t block_size_x = 1;
if (reduce_count < 512)
{
    while ((block_size_x << 1) <= reduce_count)
    {
        block_size_x <<= 1;
    }
}
else
{
    block_size_x = 512;
}
```

Reply:
nice catch, fixed